### PR TITLE
sriov pkg: Added NetworkPoolConfig

### PR DESCRIPTION
**Pull request by evgenLevin**

### CODE REVIEW

File: pkg/sriov/poolconfig.go [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+
+	srIovV1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:15]:
```suggestion
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```

Per #239

File: pkg/sriov/poolconfig.go [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+func (builder *PoolConfigBuilder) WithNodeSelector(nodeSelector map[string]string) *PoolConfigBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating SriovNetworkPoolConfig %s with Node selector: %v", builder.Definition.Name, nodeSelector)
+
+	if len(nodeSelector) == 0 {
+		builder.errorMsg = "SriovNetworkPoolConfig 'nodeSelector' cannot be empty map"
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:173]:
```suggestion
		builder.errorMsg = "SriovNetworkPoolConfig 'nodeSelector' cannot be empty map"
		
		return builder
```

File: pkg/sriov/poolconfig.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+	glog.V(100).Infof(
+		"Creating SriovNetworkPoolConfig %s with MaxUnavailable: %v", builder.Definition.Name, maxUnavailable)
+
+	if maxUnavailable.Type == intstrutil.String {
+		if strings.HasSuffix(maxUnavailable.StrVal, "%") {
+			i := strings.TrimSuffix(maxUnavailable.StrVal, "%")
+
+			v, err := strconv.Atoi(i)
+			if err != nil {
+				builder.errorMsg = fmt.Sprintf("invalid value %q: %v", maxUnavailable.StrVal, err)
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:196]:
```suggestion
				builder.errorMsg = fmt.Sprintf("invalid value %q: %v", maxUnavailable.StrVal, err)
				return builder
```

File: pkg/sriov/poolconfig.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+		}
+	} else {
+		if maxUnavailable.IntValue() < 0 {
+			builder.errorMsg = fmt.Sprintf("negative number is not allowed: %v", maxUnavailable)
+		}
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:213]: Not needed if returns are in place after setting `builder.errorMsg`.

File: pkg/sriov/poolconfig.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+		if strings.HasSuffix(maxUnavailable.StrVal, "%") {
+			i := strings.TrimSuffix(maxUnavailable.StrVal, "%")
+
+			v, err := strconv.Atoi(i)
+			if err != nil {
+				builder.errorMsg = fmt.Sprintf("invalid value %q: %v", maxUnavailable.StrVal, err)
+			}
+
+			if v > 100 || v < 1 {
+				builder.errorMsg = fmt.Sprintf("invalid value: percentage needs to be between 1 and 100: %v", maxUnavailable)
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:200]:
```suggestion
				builder.errorMsg = fmt.Sprintf("invalid value: percentage needs to be between 1 and 100: %v", maxUnavailable)
				return builder
```

File: pkg/sriov/poolconfig.go [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+			}
+		} else {
+			builder.errorMsg = fmt.Sprintf("invalid type: strings needs to be a percentage: %v", maxUnavailable)
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:203]:
```suggestion
			builder.errorMsg = fmt.Sprintf("invalid type: strings needs to be a percentage: %v", maxUnavailable)
			
			return builder
```

File: pkg/sriov/poolconfig.go [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+		}
+	} else {
+		if maxUnavailable.IntValue() < 0 {
+			builder.errorMsg = fmt.Sprintf("negative number is not allowed: %v", maxUnavailable)
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:207]:
```suggestion
			builder.errorMsg = fmt.Sprintf("negative number is not allowed: %v", maxUnavailable)
			
			return builder
```

File: pkg/sriov/poolconfig.go [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+				Name:      name,
+				Namespace: nsname,
+			},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the SriovNetworkPoolConfig is empty")
+
+		builder.errorMsg = "SriovNetworkPoolConfig 'name' cannot be empty"
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:237]:
```suggestion
		return nil, errors.New("SriovNetworkPoolConfig 'name' cannot be empty")
```

File: pkg/sriov/poolconfig.go [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the SriovNetworkPoolConfig is empty")
+
+		builder.errorMsg = "SriovNetworkPoolConfig 'namespace' cannot be empty"
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:243]:
```suggestion
		return nil, errors.New("SriovNetworkPoolConfig 'namespace' cannot be empty")
```

File: pkg/sriov/poolconfig.go [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+		apiClient: apiClient,
+		Definition: &srIovV1.SriovNetworkPoolConfig{
+			ObjectMeta: metaV1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+		}}
+
+	if name == "" {
+		builder.errorMsg = "SriovNetworkPoolConfig 'name' cannot be empty"
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:49]:
```suggestion
		builder.errorMsg = "SriovNetworkPoolConfig 'name' cannot be empty"
		
		return builder
```

File: pkg/sriov/poolconfig.go [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+	}
+
+	if nsname == "" {
+		builder.errorMsg = "SriovNetworkPoolConfig 'nsname' cannot be empty"
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:53]:
```suggestion
		builder.errorMsg = "SriovNetworkPoolConfig 'nsname' cannot be empty"
		
		return builder
```

File: pkg/sriov/poolconfig.go [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+
+	err := builder.apiClient.Delete(context.TODO(), builder.Object)
+
+	if err != nil {
+		return err
+	}
+
+	builder.Object = nil
+
+	return err
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:100]:
```suggestion
	return nil
```

File: pkg/sriov/poolconfig.go [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+
+	if err != nil {
+		glog.V(100).Infof("Failed to update SriovNetworkPoolConfig %s", builder.Definition.Name)
+
+		return nil, err
+	}
+
+	builder.Object = builder.Definition
+
+	return builder, err
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:160]:
```suggestion
	return builder, nil
```

File: pkg/sriov/poolconfig.go [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+		sriovNetworkPoolConfBuilder := &PoolConfigBuilder{
+			apiClient:  apiClient,
+			Definition: &sriovNetworkPoolConfig,
+			Object:     &sriovNetworkPoolConfig,
+		}
+
+		poolConfigBuilderObjects = append(poolConfigBuilderObjects, sriovNetworkPoolConfBuilder)
+	}
+
+	return poolConfigBuilderObjects, err
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:287]:
```suggestion
	return poolConfigBuilderObjects, nil
```

File: pkg/sriov/poolconfig.go [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+type PoolConfigBuilder struct {
+	// SriovNetworkPoolConfig definition. Used to create SriovNetworkPoolConfig object.
+	Definition *srIovV1.SriovNetworkPoolConfig
+	// Created sriovNetworkPoolConfig object.
+	Object *srIovV1.SriovNetworkPoolConfig
+	// Used in functions that define or mutate SriovNetworkPoolConfig definition.
+	// errorMsg is processed before the SriovNetworkPoolConfig object is created.
+	errorMsg string
+	// apiClient opens api connection to the cluster.
+	apiClient *clients.Settings
```
Comment (sebrandon1) [inline pkg/sriov/poolconfig.go:31]: Any chance to limit scope here?

#272 has an example.  If not that's fine.

**Review by sebrandon1 — COMMENTED** [review]
Just some comments about returns and early-returns to allow for easier unit testing in future.

**Comment by evgenLevin** [comment]
@kononovn has been added. Several unit tests have failed, but none of them are related to this PR.

File: pkg/sriov/poolconfig.go [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+	return &builder
+}
+
+// Create generates an SriovNetworkPoolConfig in the cluster and stores the created object in struct.
+func (builder *PoolConfigBuilder) Create() (*PoolConfigBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating the SriovNetworkPoolConfig %s", builder.Definition.Name)
```
Comment (kononovn) [inline pkg/sriov/poolconfig.go:70]: Please add namespace to log

File: pkg/sriov/poolconfig.go [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+	return builder, nil
+}
+
+// Delete removes an SriovNetworkPoolConfig object.
+func (builder *PoolConfigBuilder) Delete() error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	glog.V(100).Infof("Deleting the SriovNetworkPoolConfig object %s", builder.Definition.Name)
```
Comment (kononovn) [inline pkg/sriov/poolconfig.go:91]: Please add namespace to log

File: pkg/sriov/poolconfig.go [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+	return nil
+}
+
+// Exists checks whether the given SriovNetworkPoolConfig object exists in the cluster.
+func (builder *PoolConfigBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if SriovNetworkPoolConfig %s exists", builder.Definition.Name)
```
Comment (kononovn) [inline pkg/sriov/poolconfig.go:114]: Please add namespace to log

File: pkg/sriov/poolconfig_test.go [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+		testBuilder := buildValidPoolConfigTestBuilder(testSettings).WithMaxUnavailable(testCase.maxUnavailable)
+		assert.Equal(t, testBuilder.errorMsg, testCase.expectedErrorText)
+
+		if testCase.expectedErrorText == "" {
+			assert.Equal(t, testBuilder.Definition.Spec.MaxUnavailable.IntVal, testCase.maxUnavailable.IntVal)
+		}
+	}
+}
+
+func TestPullPoolConfig(t *testing.T) {
```
Comment (kononovn) [inline pkg/sriov/poolconfig_test.go:242]: You test NewPoolConfigBuilder function here instead of Pull function

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few comments above.

File: pkg/sriov/poolconfig.go [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	poolConfig := &srIovV1.SriovNetworkPoolConfig{}
+	err := builder.apiClient.Get(context.TODO(), goclient.ObjectKey{
+		Name:      builder.Definition.Name,
+		Namespace: builder.Definition.Namespace,
+	}, poolConfig)
+
+	if err != nil {
+		glog.V(100).Infof("Failed to get SriovNetworkPoolConfig %s", builder.Definition.Name)
```
Comment (kononovn) [inline pkg/sriov/poolconfig.go:143]: Please add namespace

File: pkg/sriov/poolconfig.go [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+	return poolConfig, nil
+}
+
+// Update renovates the existing SriovNetworkPoolConfig object with the new definition in builder.
+func (builder *PoolConfigBuilder) Update() (*PoolConfigBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Updating the SriovNetworkPoolConfig object %s", builder.Definition.Name)
```
Comment (kononovn) [inline pkg/sriov/poolconfig.go:157]: Please add namespace

File: pkg/sriov/poolconfig.go [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+
+	err := builder.apiClient.Update(context.TODO(), builder.Definition)
+
+	if err != nil {
+		glog.V(100).Infof("Failed to update SriovNetworkPoolConfig %s", builder.Definition.Name)
```
Comment (kononovn) [inline pkg/sriov/poolconfig.go:162]: Please add namespace

File: pkg/sriov/poolconfig.go [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+// WithNodeSelector sets nodeSelector in the SriovNetworkPoolConfig definition.
+func (builder *PoolConfigBuilder) WithNodeSelector(nodeSelector map[string]string) *PoolConfigBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating SriovNetworkPoolConfig %s with Node selector: %v", builder.Definition.Name, nodeSelector)
```
Comment (kononovn) [inline pkg/sriov/poolconfig.go:179]: Please add namespace

File: pkg/sriov/poolconfig.go [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+// WithMaxUnavailable sets MaxUnavailable in the SriovNetworkPoolConfig definition.
+func (builder *PoolConfigBuilder) WithMaxUnavailable(maxUnavailable intstrutil.IntOrString) *PoolConfigBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating SriovNetworkPoolConfig %s with MaxUnavailable: %v", builder.Definition.Name, maxUnavailable)
```
Comment (kononovn) [inline pkg/sriov/poolconfig.go:199]: Please add namespace

File: pkg/sriov/poolconfig_test.go [hunk 25]
```diff
@@ -0,0 +1,314 @@
...
+	testCases := []struct {
```
Comment (kononovn) [inline pkg/sriov/poolconfig_test.go:243]: Please add test case when api client is nil.

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few comments above.

**Review by kononovn — APPROVED** [review]
/lgtm
